engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java

```diff
@@ -0,0 +1,508 @@
+      return InvokeEqualsNodeGen.getUncached();
+    }
+
+    static InvokeEqualsNode build() {
+      return InvokeEqualsNodeGen.create();
+    }
+
+    abstract boolean execute(Atom selfAtom, Atom otherAtom);
+
+    @Specialization(guards = "cachedSelfAtomCtor == selfAtom.getConstructor()")
+    boolean invokeEqualsCachedAtomCtor(Atom selfAtom, Atom otherAtom,
+        @Cached("selfAtom.getConstructor()") AtomConstructor cachedSelfAtomCtor,
+        @Cached("getEqualsMethod(cachedSelfAtomCtor)") Function equalsMethod,
+        @Cached ExecuteCallNode executeCallNode,
+        @CachedLibrary(limit = "3") InteropLibrary interop) {
+      assert atomOverridesEquals(selfAtom);
+      Object ret = executeCallNode.executeCall(
+          equalsMethod,
+          null,
+          State.create(EnsoContext.get(this)),
+          new Object[]{selfAtom, otherAtom}
+      );
+      try {
+        return interop.asBoolean(ret);
+      } catch (UnsupportedMessageException e) {
+        throw new IllegalStateException(e);
+      }
+    }
+
+    @Specialization(replaces = "invokeEqualsCachedAtomCtor")
+    boolean invokeEqualsUncached(Atom selfAtom, Atom otherAtom,
+        @Cached ExecuteCallNode executeCallNode) {
+      Function equalsMethod = getEqualsMethod(selfAtom.getConstructor());
+      Object ret = executeCallNode.executeCall(
+          equalsMethod,
+          null,
+          State.create(EnsoContext.get(this)),
+          new Object[]{selfAtom, otherAtom}
+      );
+      assert InteropLibrary.getUncached().isBoolean(ret);
+      try {
+        return InteropLibrary.getUncached().asBoolean(ret);
+      } catch (UnsupportedMessageException e) {
+        throw new IllegalStateException(e);
+      }
+    }
+
+    @TruffleBoundary
+    static Function getEqualsMethod(AtomConstructor atomConstructor) {
+      Type atomType = atomConstructor.getType();
+      Function equalsFunction = atomConstructor
+          .getDefinitionScope()
+          .getMethods()
+          .get(atomType)
+          .get("==");
+      assert equalsFunction != null;
+      return equalsFunction;
+    }
+  }
+
+  /**
+   * Returns true if the given atom overrides `==` operator.
+   */
+  @TruffleBoundary
+  private static boolean atomOverridesEquals(Atom atom) {
+    var atomType = atom.getConstructor().getType();
+    Map<String, Function> methodsOnType = atom
+        .getConstructor()
+        .getDefinitionScope()
+        .getMethods()
+        .get(atomType);
+    if (methodsOnType != null) {
+      return methodsOnType.containsKey("==");
+    } else {
+      return false;
+    }
+  }
+
+  @Fallback
+  @TruffleBoundary
+  boolean equalsGeneric(Object left, Object right,
+      @CachedLibrary(limit = "5") InteropLibrary interop) {
+    EnsoContext ctx = EnsoContext.get(interop);
+    if (isHostObject(ctx, left) && isHostObject(ctx, right)) {
+      try {
+        return interop.asBoolean(
+            interop.invokeMember(left, "equals", right)
```

Comment:
Do we need this `equals` delegation somewhere? Is it tested?

Reply:
It is tested, e.g., in [Header_Spec.enso](https://github.com/enso-org/enso/blob/develop/test/Tests/src/Network/Http/Header_Spec.enso#L10).

And yes, we need this delegation. If you create `HashMap` via Java interop, then the following is true:
```java
interop.asBoolean(interop.invokeMember(left, "equals", right)) == true
&&
left.equals(right) == false
```
In [Equals_Spec.enso:101](https://github.com/enso-org/enso/pull/3956/files#diff-fc0aa3c55b8c9b3b56946d535160f9766a60601f665ef4cf2ec8bb249d205c19R101)